virtualenv.py

```diff
@@ -1459,6 +1459,10 @@ def install_python(home_dir, lib_dir, inc_dir, bin_dir, site_packages, clear, sy
     if sys.executable != py_executable:
         # FIXME: could I just hard link?
         executable = sys.executable
+        # If sys.executable is a redirector (for example, on Windows)
+        # get the real exe to copy using the semi-private sys._base_executable attribute)
+        if hasattr(sys, "_base_executable"):
```

Comment:
```
executable = getattr(sys,  "_base_executable", executable)
```
Simpler?

Reply:
Yeah - or even just `executable = getattr(sys, "_base_executable", sys.executable)` :-)

Reply:
Regarding docs to the "semi-private thing", I don't think there are any :-( I'm going off what @zooba said over on #1339 - well, that plus the fact that there's no other way of getting the information, so private or not we don't have much choice other than to use this attribute.

I agree, though, that it would be nice if this attribute was documented.

Reply:
Seems this is Python 3.7+ only. Link it to https://github.com/python/cpython/pull/11745 as what implemented it.

Reply:
Seems to be a site.py feature, maybe we should backport it to our site.py to make it a more general feature.

Reply:
Oh, sorry, I thought you'd followed the background in #1339. This is a new feature in Windows Python 3.7. In 3.7.2, the Python executable installed in a venv-style virtual environment changed from being a full `python.exe` to being a "redirector", that relies on being run from a venv. When virtualenv copies that redirector, it breaks. In 3.7.3, in response to some bug reports (that I haven't fully reviewed) the `sys._base_executable` attribute was added to point back to the actual exe from the source environment. That gives us a pointer to the file we need to copy.

**Unfortunately**, it looks like even this isn't sufficient - because the redirector in the intermediate venv just runs the base `python.exe`, it doesn't need the Python DLLs, and so they aren't present in the venv - so we don't copy them to the virtualenv, and as a result, `python.exe` fails because a DLL it relies on isn't present. I have no idea why this error didn't show up in my manual testing - I've just started getting errors now, even though I'm doing basically the same as I was before. Maybe I had some weird `PATH` setup that obscured the problem...

Anyway, I suspect now that this approach might be dead in the water. The whole situation is an utter mess, to be honest - and it's definitely not helped by the fact that the details aren't documented and hence aren't reliable.

I don't think it's viable to simply say that we don't support running virtualenv from within a venv - tools like pipx use venv and so running something like pew or tox from pipx will hit this limitation.

One option that *might* work to handle this is to effectively do what the `--python` option does. If we find we're running from a venv, re-invoke ourselves from the underlying "base" Python. I *think* that should be reliable, but I need to do some local tests first (as CI has Python 3.7.2, which is always going to be a problem because of the lack of `sys._base_executable`). I'll test it and update this PR once I know one way or the other.

On a somewhat related note, as you mentioned `site.py`, I'd suggest that we add support for *all* of the venv "standard" attributes (`sys.base_prefix`, `sys.base_exec_prefix`, and `sys._base_executable`) in our `site.py`, so that at least end users have a standard interface they can rely on (I don't know why venv didn't use the `sys.real_prefix` attribute that was already defined by virtualenv, but it's too late to change that now). I can prepare a separate PR to do that.

Reply:
Sounds good to me 👍 Thanks!

Reply:
Just FYI, the reason it's `_base_executable` is because it was added in 3.7.3 and isn't available in earlier versions. For 3.8, we could promote it to `base_executable`.

It should behave the same on macOS and Windows, right? Both platforms have scenarios where you can't just copy python.exe around and expect it to work, which is why the redirectors exist on both.

Reply:
It doesn't seem to be working the same in my tests (the OSX test was failing with what looked like an infinite recursion of running the same exe) but as I don't have access to an OSX machine, I can't do much other than fire speculative changes at CI. For now, I'm ignoring OSX, I'll come back to it later.

Do you know who is responsible for the redirector on OSX (I assume you are the person for the Windows one)? While it's low-level technical detail, I *really* think that this should be better documented, as at the moment it's a minefield for anyone trying to go beyond the "end user" level. I'm happy to write some docs, but I'd need to discuss the details with the people who know the reasoning behind what's going on.

Reply:
I assume it's @ned-deily, or at least he can pass the blame along to someone else.

Unfortunately, I think if all this had been properly documented, it would simply be preventing us from fixing fatal bugs in released versions, so I'm not so keen to lock in the implementation as a public interface. It's an end user level tool, and so far only virtualenv has any need to go beyond this that I'm aware of.

Reply:
>  only virtualenv has any need to go beyond this that I'm aware of

Well, certainly if virtualenv can solve the "creating one virtual environment from within another" issue, then that fixes a big chunk of the problems. But there are a lot of tools out there that for better or worse introspect the layout of virtual environments<sup>1</sup>, and that means that either the implementation just gets locked in in an ad-hoc fashion<sup>2</sup> or we need to expose/document APIs that we *can* control.

For now, though, I'm going to concentrate on trying to get running virtualenv from a venv working on Windows again<sup>3</sup> and leave the wider picture for another day.

----

<sup>1</sup> For example, a *lot* of code relies on the Python binary being `Scripts/python.exe` on Windows and `bin/python` on Unix.
<sup>2</sup> Assuming no-one is willing to break a load of longstanding code on the basis of "it's not documented so even though we didn't give you access to information you needed, it's your fault".
<sup>3</sup> And probably slightly improved, with `sys.real_prefix` pointing at the original interpreter rather than the intermediate venv.

Reply:
> Assuming no-one is willing to break a load of longstanding code on the basis of "it's not documented so even though we didn't give you access to information you needed, it's your fault"

OT: I'm super glad we bit this bullet for pip 10.